runners/samza/src/main/java/org/apache/beam/runners/samza/runtime/SamzaDoFnRunners.java

```diff
@@ -316,12 +322,25 @@ public FnDataReceiver<FnOutT> create(String pCollectionId) {
         final TimerReceiverFactory timerReceiverFactory =
             new TimerReceiverFactory(stageBundleFactory, this::timerDataConsumer, windowCoder);
 
+        Map<String, String> transformFullNameToUniqueName =
+            pTransformNodes.stream()
+                .collect(
+                    Collectors.toMap(
+                        pTransformNode -> pTransformNode.getId(),
+                        pTransformNode -> pTransformNode.getTransform().getUniqueName()));
```

Comment:
Redundant computation. Why not store this map instead of the `Collection<PipelineNode.PTransformNode>` and compute this mapping once during construction?

Reply:
Think it's okay. If we didn't want redundant computation then the above lines

```
        OutputReceiverFactory receiverFactory =
            new OutputReceiverFactory() {
              @Override
              public FnDataReceiver<FnOutT> create(String pCollectionId) {
                return (receivedElement) -> {
                  // handover to queue, do not block the grpc thread
                  outputQueue.put(KV.of(pCollectionId, receivedElement));
                };
              }
            };
        final Coder<BoundedWindow> windowCoder = windowingStrategy.getWindowFn().windowCoder();
        final TimerReceiverFactory timerReceiverFactory =
            new TimerReceiverFactory(stageBundleFactory, this::timerDataConsumer, windowCoder);
```

Should all be in the constructor?

I think the philosophy here is to keep the constructor assign only, and not contain any computation.
Code in the constructor should only look like

```
this.a = a;
this.b = b;
...
```

etc.

Reply:
Its good practice to keep constructor light weight that said, it doesn't have to always have assignments within it. Some initialization logic is okay. Other benefits doing so,
- it helps building invariants that this field isn't going to change upon construction
- handle to pTransformNode is unnecessary and can potentially be meddled with which will make this piece of non deterministic breaking the assumption that `transformFullNameToUniqueName` changes.
- Ideally, you would want only the names to be dependency injected if you don't need the `pTransformNode` at all. 

The code you pasted above is mostly blueprint and not performing any computation per-say. local variables are introduced for readability and might as well replace them with the getters of the already existing instance variables.

Reply:
By computation I mean lines that are not pure assignment.

The constructor is usually not unit tested. If we add computation logic on to it, we would want to add a unit test onto it. Ending in something like

```
class SdkHarnessDoFnRunner {
  @VisibleForTesting
  public Map<String, String> getTransformFullNameToUniqueName(){...}
}

class SdkHarnessDoFnRunnerTest {
  @Test
  public voidTestConstructor() {
     SdkHarnessDoFnRunner runner = new SdkHarnessDoFnRunner(...);
     assertEquals(runner.getTransformFullNameToUniqueName, expected);
  }
}
```

I'd rather not have extra stuff inside the constructor.

Reply:
What do you mean by "pTransformNode can be meddled with"?

Reply:
the class holds a reference of `Collection<pTransformNode>` and this reference can change over the lifecycle of this class and the transform name map that you produce changes everytime when something changes to this reference and thus causing the invariant of passing potentially different metric names upon the callsite.

By enforcing the computation to occur only once at least guarantees we always pass the same transform name maps and hence easy to reason about what the code does and how it behaves.

Reply:
The `Collection<pTransformNode>` is extracted from the `executableStage`, which is parsed from `RunnerApi.ExecutableStagePayload`.

It should be a specification of the pipeline, and not change while we run it.

If this assumption is wrong, I'll change this mapping assignment.